cpp_guidelines/.clang-format

```diff
@@ -0,0 +1,107 @@
+---
+Language: Cpp
+Standard: c++17
+DisableFormat:   false
+AccessModifierOffset: -4
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: true
+AlignConsecutiveBitFields: true
+AlignConsecutiveDeclarations: false
+AlignConsecutiveMacros: true
+AlignEscapedNewlines: Right
+AlignOperands: true
+AlignTrailingComments: true
+AllowAllArgumentsOnNextLine: false
+AllowAllConstructorInitializersOnNextLine: false
+AllowAllParametersOfDeclarationOnNextLine: false
+AllowShortBlocksOnASingleLine: Never
+AllowShortCaseLabelsOnASingleLine: false
+AllowShortEnumsOnASingleLine: false
+AllowShortFunctionsOnASingleLine: None
+AllowShortIfStatementsOnASingleLine: Never
+AllowShortLambdasOnASingleLine: All
+AllowShortLoopsOnASingleLine: false
+AlwaysBreakAfterReturnType: None
+AlwaysBreakBeforeMultilineStrings: false
+AlwaysBreakTemplateDeclarations: Yes
+BinPackArguments: false
+BinPackParameters: false
+BitFieldColonSpacing: Both
+BreakBeforeBinaryOperators: All
+BreakBeforeBraces: Allman
```

Comment:
If this is about east and west coast style, I vote for west coast because it is much easier for the eye to see where a codeblock is opened and closed.

Reply:
No, this is about
```
void f() {
  if ()
  {
  }
  else
  {
  }
}
```
vs
```
void f() {
  if () {
  } else {
  }
}
```

East and west const are not yet handled by clang-format. They are working on that: https://reviews.llvm.org/D69764

Reply:
It turns out, this also affects empty constructors (Allman):
```
    template <std::size_t Dim>
    struct UserDomainCoordRange
    {
        UserDomainCoordRange(UserDomain<Dim> size) : size(size)
        {
        }
    };
```
vs. custom brace wrapping with `SplitEmptyFunction: false` and `AllowShortBlocksOnASingleLine: Empty`
```
    template <std::size_t Dim>
    struct UserDomainCoordRange
    {
        UserDomainCoordRange(UserDomain<Dim> size) : size(size) {}
    };
```

Reply:
**VOTE**
current (Allman) | Attach | Custom|
--------|-------------|---|
 0      |   1         | 0 |

I would like to keep all opening braces on the same line to save vertical screen space. There is also the Custom option that allows finer settings.

Reply:
current (Allman) | Attach | Custom
-- | -- | --
1 | 1 | 0

I am especially against attaching *all* braces (i.e. "Java-style": attaching braces for functions and classes). I am less strongly against it for staements (i.e. "K&R").

Reply:
current (Allman) | Attach | Custom
-- | -- | --
2 | 1 | 0

I like the vertical structure in the code.

Here is a **negative** example from *alpaka*  why I'am against it.

```C++
            struct BlockSharedMemDynSizeBytes<BlockSyncTestKernel, TAcc> {
                //-----------------------------------------------------------------------------
                //! \return The size of the shared memory allocated for a
                //! block.
                template <typename TVec>
                ALPAKA_FN_HOST_ACC static auto getBlockSharedMemDynSizeBytes(
                    BlockSyncTestKernel const& blockSharedMemDyn,
                    TVec const& blockThreadExtent,
                    TVec const& threadElemExtent,
                    bool* success) -> idx::Idx<TAcc> {
                    using Idx = alpaka::idx::Idx<TAcc>;

                    alpaka::ignore_unused(blockSharedMemDyn);
                    alpaka::ignore_unused(threadElemExtent);
                    alpaka::ignore_unused(success);
                    return static_cast<idx::Idx<TAcc>>(sizeof(Idx))
                           * blockThreadExtent.prod();
                }
            };
```

Reply:
current (Allman) | Attach | Custom
-- | -- | --
3 | 1 | 0

Reply:
current (Allman) | Attach | Custom
--------|-----------|-------------|
 4      |   1 | 0

Reply:
current (Allman) | Attach | Custom
-- | -- | --
5 | 1 | 0

Reply:
current (Allman) | Attach | Custom
-- | -- | --
6 | 1 | 0

Reply:
current (Allman) | Attach | Custom
-- | -- | --
6 | 2 | 0

Attaching the {} would be consistent with our style for template braces <> and tuple braces ()..

Reply:
Reminder: Voting will be closed on 2020-09-26 at 13:44 HZDR time.

Reply:
**VOTE**
current (Allman) | Attach | Custom |
--------|-----------|------|
 6      |   3    | 0 |

Reply:
Voting closed, the winner is `Allman`.